File: Data Structures & Algorithms/depth-of-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        # if not root:
        #     return 0
        
        # return 1+ max(self.maxDepth(root.left),self.maxDepth(root.right))

        # q = deque() dfs/level order

        # if root:
        #     q.append(root)
        # level=0
        # while q:
        #     for i in range(len(q)):
        #         node = q.popleft()
        #         if node.left:
        #             q.append(node.left)
        #         if node.right:
        #             q.append(node.right)
        #     level+=1
        # return level

        if not root:
            return 0
        
        res=0
        stack= [[root,1]]
        cur = root

        while stack:
            node,v = stack.pop()
            if node:
                res = max(res,v)
                stack.append([node.left,1+v])
                stack.append([node.right,1+v])
        return res

File: Data Structures & Algorithms/house-robber-iii/submission-3.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
func rob(root *TreeNode) int {
    result := dfs(root)
    return max(result[0], result[1])
}

func dfs(root * TreeNode) [2]int {
    if root == nil{
        return [2]int{0, 0}
    }

    left := dfs(root.Left)
    right := dfs(root.Right)

    withRoot := root.Val + left[1] + right[1]
    withoutRoot := max(left[0], left[1]) + max(right[0], right[1])

    return [2]int{withRoot, withoutRoot} 
}

func max(a, b int) int{
    if a > b{
        return a
    }
    return b
}